### ISSUE TITLE
iOS Location services permission requested again after selecting "When in use"

### ISSUE THREAD

**Issue by irikuo** [transistorsoft/cordova-background-geolocation-lt#494]
## Your Environment
* Plugin version: 2.8.5
* Platform: iOS
* OS version: 11 (I think it happens in 10 as well)
* Device manufacturer / model: iPhone 7
* Cordova version (`cordova -v`): 6.3.1
* Cordova platform version (`cordova platform ls`): ios 4.1.1
* Plugin config
```
locationAuthorizationRequest: 'Always',
locationAuthorizationAlert: {
    titleWhenNotEnabled: "Location-services not enabled",
    titleWhenOff: "Turn on Location Services",
    instructions: "Enable 'Always' in Location Services for this app in order to view your location",
    cancelButton: "Cancel",
    settingsButton: "Settings"
},
```
## Expected Behavior
I'd expect to be able to make a location permission optional. I explain this: if `locationAuthorizationRequest` is set to `'Always'` and the user chooses something else (either reject the permission or selecting `When in use`), it shouldn't display the dialog to enable the location services again, the user just made a decision. 

## Actual Behavior
When the user selects the reject option, the dialog goes away, that's fine. But if the user chose "When in use", there's a popup to enable location services again that shouldn't be displayed.

## Steps to Reproduce
1. Configure `locationAuthorizationRequest` to `Always`.
2. Trigger the permission request somewhere in your app.
3. Select "When in use".
4. A popup opens requesting the location services again.

**Comment by christocracy** [comment]
Listen to providerchange event to detect when user changes permission.

 locationAuthorizationRequest represents the value that the **plugin should EXPECT** the user to have selected.  If **you** want to make this optional, listen to providerchange event and execute #setConfig as desired, if you wish to tell the plugin to expect some other locationAuthorizationRequest.

There’s a **major** change in behavior when the user selects WhenInUse.  Much of plugin’s features will no longer work, including geofences and Motion-detection (ie automatically engaging tracking when device is detected to be moving).

**Comment by irikuo** [comment]
Hi @christocracy! 

Thanks for the reply. What I mean is, why if the user rejects the permission there's no popup asking again for it, but if the user chooses "When in use" the popup to ask for the permission shows again? From my point of view, in both cases the app doesn't get enough permissions to work, so I don't really get the difference in behaviour. I don't know if I'm explaining this very well (not a native speaker here, sorry).

> There’s a major change in behavior when the user selects WhenInUse. Much of plugin’s features will no longer work, including geofences and Motion-detection (ie automatically engaging tracking when device is detected to be moving).

I know this, that's exactly why I believed the behaviour when the requested permission is "Always" might be the same selecting "When in use" or "Deny".

**Comment by christocracy** [comment]
There *will* be another popup when a location is requested (eg: `#getCurrentPosition`, `#start`)

I suggest you start by implementing a `providerchange` event-handler.  See the docs.

**Comment by irikuo** [comment]
Sorry, I forgot to mention. I have a `providerchange` event that stops the plugin in case it has no permissions. But even doing that, you can see the popup briefly after selecting "When in use" (when the `providerchange` event is triggered, the popup disappears automatically). We think that might be confusing for the user.

It starts to feel as I'm missing something here... I still don't fully understand why "When in use" and "Deny" behave in a different way when the requested permission is "Always". Am I suggesting something weird here? Just trying to get the behaviour of the plugin and why it works like it does :)

**Comment by christocracy** [comment]
Show me your `providerchange` handler

**Comment by irikuo** [comment]
This is it:

```
bgGeo.on('providerchange', function(provider) {
  // If the location permissions are disabled or different than "Always", stop the plugin
  if (!provider.enabled || provider.status != 3) {
    bgGeo.stop();
  }
});
```

**Comment by christocracy** [comment]
If you want the user to be able select *either* `Always` or `WhenInUse`, you must tell the plugin which `locationAuthorizationRequest` to *expect* by using `#setConfig`
```javascript
BackgroundGeolocation.on('providerchange', (status) => {
  switch(provider.status) {
    case BackgroundGeolocation.AUTHORIZATION_STATUS_DENIED:
      // Android & iOS
      console.log('- Location authorization denied');
      break;
    case BackgroundGeolocation.AUTHORIZATION_STATUS_ALWAYS:
      // Android & iOS
      console.log('- Location always granted');
      BackgroundGeolocation.setConfig({locationAuthorizationRequest: 'Always'});
      break;
    case BackgroundGeolocation.AUTHORIZATION_STATUS_WHEN_IN_USE:
      // iOS only
      console.log('- Location WhenInUse granted');
      BackgroundGeolocation.setConfig({locationAuthorizationRequest: 'WhenInUse'});
      break;
  }
});
```

`locationAuthorizationRequest` tells the plugin which value **to expect**.  You need to use `#setConfig` to change the plugin's *expectations*.

**Comment by irikuo** [comment]
Sorry, I feel like I am not comunicating the issue properly. I want the user just to select `Always`, and I am setting that using:

```
BackgroundGeolocation.configure({
    locationAuthorizationRequest: 'Always',
    ... // more config
});
```

so if they choose `WhenInUse` or `Deny`, I thought the behavior would be the same, but it's not  what I am getting. What I am getting is:

The user selects `Deny` -> The dialog goes away, that's good and what I expect.
The user selects `WhenInUse` -> The dialog goes away, but another popup is displayed to request the location permissions. If I use the `'providerchange'` handler, I can make it disappear by calling `.stop`, but it's still flashed for a moment to the user, and that's confusing (it's a popup that shows and closes automatically and they cannot even read it).

To sum up: I think if I define `'Alsways'` as the requested permission, anything that it's not that value should behave in the same way (as you said `'WhenInUse'` involves huge changes on the behavior.

I hope now it's clearer what I mean :)

**Comment by christocracy** [comment]
It seems iOS 11 has made a change here.  The OS *used to* awaken a suspended app to receive the location-authorization-change callback, so that by the time the user flipped back to their app from the Settings screen, the Javascript `providerchange` event would have already fired and the `#stop` method would already have been executed, hiding the authorization popup.

It seems that iOS 11 now runs the location-authorization-change callback only after the app comes to foreground.

I have solved this by simply adding a short delay to showing the popup, giving your `providerchange` callback the chance to execute `#stop` within this delay.

This change is available now in the Premium Github repo, available to licensed customers.  It will be posted here in the public repo when I next deploy an update.  Could be a few more weeks.

**Comment by irikuo** [comment]
Cool! That sounds good @christocracy, many thanks :)

**Comment by christocracy** [comment]
Try `2.9.1`
